[SMST] feat: Use compact SMST proofs

red-0ne opened the pull request:
## Summary

This PR changes the proofs being sent in `MsgSubmitPRoof` and stored by the `SubmitProof` keeper from `SparseMerkleClosestProof` to `SparseCompactMerkleClosestProof`.

## Issue

`Proofs` one of the most block-space consuming primitives. The SMST support compact proofs that could help reduce their sizes.

- #758 

## Type of change

Select one or more from the following:

- [x] New feature, functionality or library
- [x] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code

Olshansk reviewed (approved):
Amazed that everyone just worked!

Any chance you can add a log or a unit test to show the difference in size?

I'm REALLy curious and can be done in a followup.

red-0ne commented:
> Any chance you can add a log or a unit test to show the difference in size?

@Olshansk, I'm really grateful that you called this out.

I added this test [/pkg/relayer/session/sessiontree_test.go](https://github.com/pokt-network/poktroll/pull/823/files#diff-d13257bc0fb6f968fa7ca4f63a57cb899a302c9a5f22d63f87a67f3f708014c2)

Which shows that there is barely any improvement in the proof size :cry: 

I added a (gzip) compressed proofs to the test to have something to compare with.

Running for different leaf sizes, 1000 run each and accumulating size, this a sample result

```
numLeaf=10: compactionRatio: 0.990917, compressionRatio: 0.804937
numLeaf=100: compactionRatio: 0.989091, compressionRatio: 0.803551
numLeaf=1000: compactionRatio: 0.992390, compressionRatio: 0.805643
numLeaf=10000: compactionRatio: 0.989311, compressionRatio: 0.803930
numLeaf=100000: compactionRatio: 0.992038, compressionRatio: 0.805471
numLeaf=1000000: compactionRatio: 0.990228, compressionRatio: 0.804995
```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
```
#PUC to link to the results: https://github.com/pokt-network/poktroll/pull/823#issuecomment-2363987920

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+		cumulativeProofSize := 0
+		cumulativeCompactProofSize := 0
+		cumulativeGzippedProofSize := 0
+		for numLeaf := 0; numLeaf <= 1000; numLeaf++ {
```
We're using the same variable `numLeaf` internally and externally. Please fix.

Olshansk replied:
Fix the naming for booth loop variables

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+	// Run the test for different number of leaves.
+	for numLeaf := 10; numLeaf <= 1000000; numLeaf *= 10 {
+		// We run the test 1000 times for each number of leaves to remove the randomness bias.
+		cumulativeProofSize := 0
+		cumulativeCompactProofSize := 0
+		cumulativeGzippedProofSize := 0
```
```suggestion
		cumulativeProofSize := 0
		cumulativeCompactProofSize := 0
		cumulativeGzippedProofSize := 0
		// We run the test 1000 times for each number of leaves to remove the randomness bias.
```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+
+func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+	// Run the test for different number of leaves.
+	for numLeaf := 10; numLeaf <= 1000000; numLeaf *= 10 {
```
```suggestion
	for numLeafs := 10; numLeafs <= 1000000; numLeafs *= 10 {
```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+			proofBz, err := proof.Marshal()
+			require.NoError(t, err)
+
+			// Accumulate the proof size over 1000 runs.
```
Don't put numbers like this in the comments.

If you (or another developer) ever update the loop, there's no chance you'll update the comment.

Comments should be "ever green".

If they're not "ever green", they should involve a TODO.

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+			compactProofBz, err := compactProof.Marshal()
+			require.NoError(t, err)
+
+			// Accumulate the compact proof size over 1000 runs.
```
Ditto

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -1,3 +1,96 @@
+			// Accumulate the gzipped proof size over 1000 runs.
+			cumulativeGzippedProofSize += len(buf.Bytes())
+
+			//t.Logf(
```
Either delete or merge in the things that we need

Olshansk commented:
@red-0ne Thanks for putting in the extra test. This is fine and what I expected.

Want to use this as a learning lesson. Nothing we don't already know but something I want to reiterate:

1. `Compact proof` sounds sexy, eh? This is a classic example of "sounds sexy to an engineer" but not actually useful.
     - Probabilistic proof is what will actually ensure scalability
2. Understanding what's happening under the hood
     - We inherited "Compact Proofs" from the library we forked
     - I didn't implement it, neither did Harry, neither did you
     - None of us looked in the source code
     - I'm 100% sure we can make it more efficient, but again, it "sounds sexy", right?
3. The importance of visibility / observability
    - This doesn't always need to be a dashboard
    - It doesn't always need to be telemetry
    - It could be a log or a unit test
    - This is a great example of using `t.Log`
    - This raises the importance of a dashboard for probabilistic proofs

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -12,23 +12,34 @@ import (
 )
 
+const (
+	maxLeafs      = 10000
```
#PUC details for both of these constants

Olshansk commented on x/proof/keeper/proof_validation_test.go:
```diff
@@ -611,8 +611,10 @@ func TestEnsureValidProof_Error(t *testing.T) {
-				err = sparseMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof)
+				err = sparseCompactMerkleClosestProof.Unmarshal(proof.ClosestMerkleProof)
+				require.NoError(t, err)
+				var sparseMerkleClosestProof *smt.SparseMerkleClosestProof
+				sparseMerkleClosestProof, err = smt.DecompactClosestProof(sparseCompactMerkleClosestProof, &protocol.SmtSpec)
 				require.NoError(t, err)
```
```suggestion

				var sparseMerkleClosestProof *smt.SparseMerkleClosestProof
				sparseMerkleClosestProof, err = smt.DecompactClosestProof(sparseCompactMerkleClosestProof, &protocol.SmtSpec)
				require.NoError(t, err)

```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
-		//	"numLeaf=%d: cumulativeProofSize: %d, cumulativeCompactProofSize: %d, cumulativeGzippedProofSize: %d",
-		//	numLeaf, cumulativeProofSize, cumulativeCompactProofSize, cumulativeGzippedProofSize,
-		//)
+		compactionRatio := float32(cumulativeCompactProofSize) / float32(cumulativeProofSize)
+		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
```
```suggestion
		// Calculate how much more efficient compact SMT proofs are
		compactionRatio := float32(cumulativeCompactProofSize) / float32(cumulativeProofSize)

		// Calculate how much more efficient gzipped SMT proofs are
		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+		compactionRatio := float32(cumulativeCompactProofSize) / float32(cumulativeProofSize)
+		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
+
+		// Gzip compression is more efficient than compaction.
```
```suggestion
		// Gzip compression is more efficient than SMT compaction.
```

Olshansk commented on pkg/relayer/session/sessiontree_test.go:
```diff
@@ -68,29 +79,19 @@ func TestSessionTree_CompactProofsAreSmallerThanNonCompactProofs(t *testing.T) {
+		compressionRatio := float32(cumulativeGzippedProofSize) / float32(cumulativeProofSize)
+
+		// Gzip compression is more efficient than compaction.
+		require.Greater(t, compactionRatio, compressionRatio)
```
This seems to contradict the comment above, no?

red-0ne replied:
Inverted how the ratio is calculated, now the higher the better.

h5law commented:
> @red-0ne Thanks for putting in the extra test. This is fine and what I expected.
> 
> Want to use this as a learning lesson. Nothing we don't already know but something I want to reiterate:
> 
> 1. `Compact proof` sounds sexy, eh? This is a classic example of "sounds sexy to an engineer" but not actually useful.
>    
>    * Probabilistic proof is what will actually ensure scalability
> 2. Understanding what's happening under the hood
>    
>    * We inherited "Compact Proofs" from the library we forked
>    * I didn't implement it, neither did Harry, neither did you
>    * None of us looked in the source code
>    * I'm 100% sure we can make it more efficient, but again, it "sounds sexy", right?
> 3. The importance of visibility / observability
>    
>    * This doesn't always need to be a dashboard
>    * It doesn't always need to be telemetry
>    * It could be a log or a unit test
>    * This is a great example of using `t.Log`
>    * This raises the importance of a dashboard for probabilistic proofs

Little late to this but just going through my notifications. I actually did implement the CompactClosestProof type to fit the pattern with the other proof types. Had some brief ideas about how they could be compacted properly (was in a rush when implementing this):

1. Clip the `Path` field to `Depth` bits
2. Replace the `FlippedBits` byte matrix with a bit mask where `1`s represent bits flipped, the `[][]byte` can be replaced with `uint{Depth}` as each bit represents an index (up to 4 `uint64`s for a full 256-bit worst case encoding)
3. Instead of encoding the `Depth` as the `[]byte` form of the original `int` use a single `byte` (this assumes a 256 bit hasher or figure out a 9-bit type to generalise and include 512-bit hashers)

Realisticallly this method is super optimisable, I didn't really try to optimise it as it was always don't pre-optimise. I'm sure you guys will think of even more, actually doing compression on the encoded bits would be a interesting idea once they are encoded better.
